include/deal.II/numerics/vector_tools.templates.h

```diff
@@ -891,6 +895,267 @@ namespace VectorTools
+                                  const DoFHandler<dim, spacedim>                           &dof,
+                                  const ConstraintMatrix                                    &constraints,
+                                  const Quadrature<dim>                                     &quadrature,
+                                  const Function<spacedim, typename VectorType::value_type> &function,
+                                  VectorType                                                &vec_result,
+                                  const bool                                                 enforce_zero_boundary,
+                                  const Quadrature<dim-1>                                   &q_boundary,
+                                  const bool                                                 project_to_boundary_first)
+    {
+      switch (dof.get_fe().degree)
+        {
+        case 1:
+          dealii::VectorTools::project_generic<components, 1>
+          (mapping, dof, constraints, quadrature, function, vec_result,
+           enforce_zero_boundary, q_boundary, project_to_boundary_first);
+          break;
+
+        case 2:
+          dealii::VectorTools::project_generic<components, 2>
+          (mapping, dof, constraints, quadrature, function, vec_result,
+           enforce_zero_boundary, q_boundary, project_to_boundary_first);
+          break;
+
+        case 3:
+          dealii::VectorTools::project_generic<components, 3>
+          (mapping, dof, constraints, quadrature, function, vec_result,
+           enforce_zero_boundary, q_boundary, project_to_boundary_first);
+          break;
+
+        case 4:
+          dealii::VectorTools::project_generic<components, 4>
+          (mapping, dof, constraints, quadrature, function, vec_result,
+           enforce_zero_boundary, q_boundary, project_to_boundary_first);
+          break;
+
+        case 5:
+          dealii::VectorTools::project_generic<components, 5>
+          (mapping, dof, constraints, quadrature, function, vec_result,
+           enforce_zero_boundary, q_boundary, project_to_boundary_first);
+          break;
+
+        case 6:
+          dealii::VectorTools::project_generic<components, 6>
+          (mapping, dof, constraints, quadrature, function, vec_result,
+           enforce_zero_boundary, q_boundary, project_to_boundary_first);
+          break;
+
+        case 7:
+          dealii::VectorTools::project_generic<components, 7>
+          (mapping, dof, constraints, quadrature, function, vec_result,
+           enforce_zero_boundary, q_boundary, project_to_boundary_first);
+          break;
+
+        case 8:
+          dealii::VectorTools::project_generic<components, 8>
+          (mapping, dof, constraints, quadrature, function, vec_result,
+           enforce_zero_boundary, q_boundary, project_to_boundary_first);
+          break;
+
+        default:
+          Assert(false, ExcNotImplemented());
```

Comment:
Actually, I neither check for the number of components nor for the degree of the `FiniteElement` with `use_matrix_free`, but I see the point in changing this.